### PR TITLE
Noc description support and modelling in VPR

### PR DESCRIPTION
**Pull request by Srivat97**
Extended the FPGA architecture description to now include NoC specific information. Then updated VPR to process the NoC information provided by the user and create an internal model. Finally, the graphics was updated to display the NoC to the user.

#### Description

libarchfpga
- Updated the xml parser code to now include the additional tags used to describe the NoC. This information is then stored internally.

libvpr
- Added a new class called NocRouter which represents a router within the NoC
- Added a new class called NocLink which represents a link within the NoC (connections between routers)
- Added a new class called NocStorage which represents the entire NoC (list of routers and links that together make up the NoC)
- Created a new context called noc_ctx, this contains all the NoC related material so that it can be accessed globally
- Added a new function called setup_noc that is called when the device is being created. This is responsible for building the NoC based on the information provided in the FPGA architecture description file. This includes creating all the routers and links.
- Added a new set of command line options for the NoC.
- Added a new button to display the NoC (draws the routers and links). Also can draw the NoC link usage as a heat map.

#### Motivation and Context

The eventual goal is to have a CAD tool that is capable of targeting FPGA architectures with an embedded NoC. But first we need to be able to get specifications about the NoC, such as its topology, location on the FPGA, and any design constraints (max link bandwidth, link latency, router latency and etc..). Once we have the NoC specifications, we need to model it so that we can ensure any steps taken in the CAD flow always meet the NoC design specifications. So the initial steps of getting the NoC information from the user and modelling it as done in this pull request.

#### How Has This Been Tested?

- Unit tests were added to verify the newly developed functions
- Additionally, sample FPGA architecture description files were created that included a description of an embedded NoC. VPR was run on the sample FPGA architecture description files with arbitrary designs and the NoC was displayed visually to ensure that the new feature worked correctly.  

#### Types of changes

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by vaughnbetz** [comment]
Should add an architecture file to test the overall flow works (augments the unit tests).  It's OK if it's a simple architecture-in-progress.

File: vpr/src/base/noc_link.cpp
Comment (vaughnbetz) [inline vpr/src/base/noc_link.cpp:1]: Suggest making this doxygen compatible. Then you can make a NoC subcategory in the VPR API documentation on the VTR website.

File: vpr/src/base/noc_link.cpp
Comment (vaughnbetz) [inline vpr/src/base/noc_link.cpp:44]: Would get_num_traffic_flows be a better name?
get_num_ is a bit shorter and probably just as clear.

File: vpr/src/base/noc_link.h
Comment (vaughnbetz) [inline vpr/src/base/noc_link.h:6]: This is where I meant Doxygen!

File: vpr/src/base/noc_link.h
Comment (vaughnbetz) [inline vpr/src/base/noc_link.h:29]: num_traffic_flows won't be enough --> will need currentTraffic and maybe back ids to the traffic flows that are going through here.

File: vpr/src/base/noc_router.cpp
Comment (vaughnbetz) [inline vpr/src/base/noc_router.cpp:2]: Might want to say "physical router"

File: vpr/src/base/noc_storage.cpp
Comment (vaughnbetz) [inline vpr/src/base/noc_storage.cpp:6]: Same idea; most description should go in .h file.

File: vpr/src/base/noc_storage.cpp
Comment (vaughnbetz) [inline vpr/src/base/noc_storage.cpp:62]: Suggest returning references or const references (your call) to the noc_links and noc_routers so the end caller can just call their functions directly. Otherwise you have to wrap each method in the lower level object with one here in this composite object, which is likely to get tedious.

**Review by vaughnbetz — CHANGES_REQUESTED** [review]
Good overall, but some changes requested.
Most important: unit tests must test the real code (not a copy).

File: libs/libarchfpga/src/arch_util.cpp
Comment (Srivat97) [reply libs/libarchfpga/src/arch_util.cpp:208]: fixed

File: libs/libarchfpga/src/physical_types.h
Comment (Srivat97) [reply libs/libarchfpga/src/physical_types.h:1778]: done

File: libs/libarchfpga/src/physical_types.h
Comment (Srivat97) [reply libs/libarchfpga/src/physical_types.h:1789]: done

File: libs/libarchfpga/src/physical_types.h
Comment (Srivat97) [reply libs/libarchfpga/src/physical_types.h:1796]: done

File: libs/libarchfpga/src/read_xml_arch_file.cpp
Comment (Srivat97) [reply libs/libarchfpga/src/read_xml_arch_file.cpp:4671]: fixed

File: vpr/src/base/noc_link.cpp
Comment (Srivat97) [reply vpr/src/base/noc_link.cpp:1]: fixed

File: vpr/src/base/noc_link.cpp
Comment (Srivat97) [reply vpr/src/base/noc_link.cpp:44]: fixed

File: vpr/src/base/noc_link.h
Comment (Srivat97) [reply vpr/src/base/noc_link.h:6]: fixed

File: vpr/src/base/noc_router.cpp
Comment (Srivat97) [reply vpr/src/base/noc_router.cpp:7]: Fixed and moved comments to header file

File: vpr/src/base/noc_router.cpp
Comment (Srivat97) [reply vpr/src/base/noc_router.cpp:2]: Fixed

File: vpr/src/base/noc_router.cpp
Comment (Srivat97) [reply vpr/src/base/noc_router.cpp:53]: Fixed

File: vpr/src/base/noc_router.h
Comment (Srivat97) [reply vpr/src/base/noc_router.h:7]: Fixed

File: vpr/src/base/noc_router.h
Comment (Srivat97) [reply vpr/src/base/noc_router.h:28]: Fixed

File: vpr/src/base/noc_storage.cpp
Comment (Srivat97) [reply vpr/src/base/noc_storage.cpp:6]: Fixed

File: vpr/src/base/noc_storage.cpp
Comment (Srivat97) [reply vpr/src/base/noc_storage.cpp:79]: Fixed

File: vpr/src/base/noc_storage.cpp
Comment (Srivat97) [reply vpr/src/base/noc_storage.cpp:156]: Changed the router "id" variable to "user_id"

File: vpr/src/base/noc_storage.cpp
Comment (Srivat97) [reply vpr/src/base/noc_storage.cpp:180]: fixed

File: vpr/src/base/setup_noc.cpp
Comment (Srivat97) [reply vpr/src/base/setup_noc.cpp:169]: changed

File: vpr/src/base/vpr_context.h
Comment (Srivat97) [reply vpr/src/base/vpr_context.h:408]: This makes sense. Moved to the NocStorage class

File: vpr/test/test_setup_noc.cpp
Comment (Srivat97) [reply vpr/test/test_setup_noc.cpp:13]: Done

File: vpr/src/draw/draw.cpp
Comment (Srivat97) [reply vpr/src/draw/draw.cpp:193]: Can I change some functions that I use externally from being static?

Comment (vaughnbetz) [reply vpr/src/draw/draw.cpp:193]: Sure.  There's nothing too special about static; just a way to hide names inside a file if they aren't needed outside.  If they are needed outside, feel free to remove the static and put their declarations in the .h file.

**Comment by vaughnbetz** [comment]
Infrstructure issue; CI should pass once I merge Alireza's change.
Good to get this merged soon; draw will have to be updated as it was heavily refactored.

File: vpr/src/draw/draw_noc.cpp
Comment (Srivat97) [reply vpr/src/draw/draw_noc.cpp:359]: Yes, it was a dot product not a cross product. Changed and included comments.

File: vpr/src/draw/draw_noc.cpp
Comment (Srivat97) [reply vpr/src/draw/draw_noc.cpp]: Done

File: vpr/src/draw/draw_noc.cpp
Comment (Srivat97) [reply vpr/src/draw/draw_noc.cpp:457]: Will leave as is for now. I'll think of a better solution and change later.

**Comment by Srivat97** [comment]
@vaughnbetz This PR should be ready to merge. Fixed all requested changes or made a comment to explain why I didn't make the change. 

I also added some sample architecture files I used to test this new feature. They can be found under "$VTR_ROOT/vtr_flow/arch/noc". There are two architecture files in there, one has a mesh topology and the other has a star topology. I also updated the README found in "$VTR_ROOT/vtr_flow/" to help new users find these files and explain their purpose.

**Comment by vaughnbetz** [comment]
Looks good -- thanks!

**Pull request merged**
